File: app/src/main/java/com/example/android/architecture/services/PostService.java
package com.example.android.architecture.services;

import com.example.android.architecture.data.remote.apis.PostApi;
import com.example.android.architecture.models.Post;
import com.google.gson.Gson;
import io.reactivex.Flowable;
import java.util.List;
import java.util.Map;

/**
 * Created by YaoHaitao on 2018/3/8.
 */

public class PostService {

  private final PostApi mPostApi;

  public PostService(PostApi postApi) {
    this.mPostApi = postApi;
  }

  public Flowable<List<Post>> fetchPosts() {
    return mPostApi.fetchPosts();
  }

  public Flowable<Post> fetchPostById(int postId) {
    return mPostApi.fetchPostByPostId(postId);
  }

  public Flowable<Post> insertPost(Post post) {
    Gson gson = new Gson();
    Map<String, Object> postMap = gson.fromJson(gson.toJson(post), Map.class);
    return mPostApi.insertPost(postMap);
  }

  public Flowable<Post> updatePost(Post post) {
    Gson gson = new Gson();
    Map<String, Object> postMap = gson.fromJson(gson.toJson(post), Map.class);
    return mPostApi.updatePost(postMap);
  }

  public Flowable<Post> deletePostByPostId(int postId) {
    return mPostApi.deletePostByPostId(postId);
  }
}

File: app/src/main/java/com/example/android/architecture/services/PostService.kt
package com.example.android.architecture.services

import com.example.android.architecture.data.remote.PostApi
import com.example.android.architecture.models.Post
import com.google.gson.Gson

/**
 * ポストサービス
 * ポストに関する操作はここで実装される。
 */
class PostService constructor(val postApi: PostApi) {

  fun fetchPosts() = postApi.fetchPosts()

  fun fetchPost(id: Int) = postApi.fetchPost(id)

  fun createPost(post: Post) = {
    val gson = Gson()
    val postMap = gson.fromJson<Map<String, Any?>>(gson.toJson(post))
//    val postMap: Map<String, Any?> = gson.fromJson(gson.toJson(post), Map::class.java) as Map<String, Any?>
    postApi.createPost(postMap)
  }()

  fun updatePost(post: Post) = {
    val gson = Gson()
    val postMap = gson.fromJson<Map<String, Any?>>(gson.toJson(post))
//    val postMap: Map<String, Any?> = gson.fromJson(gson.toJson(post), Map::class.java) as Map<String, Any?>
    postApi.updatePost(postMap)
  }()

  fun deletePost(id: Int) = postApi.deletePost(id)

  private inline fun <reified T> Gson.fromJson(json: String): T {
    return fromJson(json, T::class.java)
  }
}